dcrec/secp256k1/sigerror.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package secp256k1
+
+import (
+	"fmt"
+)
+
+// SignatureErrorCode identifies a kind of signature error.  It has full support
+// for errors.Is and errors.As, so the caller can directly check against an
+// error code when determining the reason for an error.
+type SignatureErrorCode int
+
+// These constants are used to identify a specific SignatureError.
+const (
+	// ErrSigTooShort is returned when a signature that should be a DER
+	// signature is too short.
+	ErrSigTooShort SignatureErrorCode = iota
```

Comment:
I suggest starting these at 1 so that the zero value can mean no error.

Reply:
Not necessarily if you have the error code in a struct.

Reply:
I can update it, but I have to say that it really doesn't make much sense to me.  Why would you ever produce an error when there wasn't an error?

Reply:
I should note that I totally understand in languages like C that don't have multiple returns it is common practice to have such sentinel values, but that seems to explicitly be against the Go  ethos and is even called out in Effective Go as something to be avoided iirc.

Reply:
A struct may contain different error codes and conditions besides just this error kind.

Also, if this error was being serialized across the wire, I would expect it to be sent as a plain integer, and reading into a error code variable, with a check for zero to indicate that there was no error.

Reply:
Might be worthwhile in a separate PR.  After looking at it, I really don't want to add it in this one as it will complicate the error handling since there would be special cases that have to be dealt with in various places.